---
int8 num_samples    